### DISCUSSION TITLE
JupyterLab, TypeError: getOwnMetadataKeys is not a function

### DISCUSSION THREAD

**Discussion by manoj30075** [inversify/InversifyJS#1567] [Q&A]
I’m currently developing an extension for JupyterLab. As part of this process, I’m utilizing the InversifyJS library from npm. To use it, I have to import ‘reflect-metadata’ in my index.ts file, which I have done. However, when I create an injectable class and use it in my extension, I encounter the following error:

```
TypeError: getOwnMetadataKeys is not a function
    at Object.isProviderFor (Reflect.js:1135:1)
    at getProviderNoCache (Reflect.js:941:1)
    at Object.getProvider (Reflect.js:955:1)
    at GetMetadataProvider (Reflect.js:1158:1)
    at OrdinaryDefineOwnMetadata (Reflect.js:611:1)
    at decorator (Reflect.js:195:1)
    at DecorateConstructor (Reflect.js:549:1)
    at Reflect.decorate (Reflect.js:143:1)
    at push.../shared/components/localTelemetry.ts.__decorate (icons.ts:4:72)
    at ../shared/components/localTelemetry.ts (localTelemetry.ts:7:28)
```

This is part of my code

```
@injectable()
export class LocalTelemetry implements TelemetryService {
  constructor(
    @inject(NAMES.customerId)
    private customerId: string = '',
    @inject(NAMES.outputChannelTelemetry)
    private outputChannel: Logger
  ) {}
```
And I have imported the reflect-metadata at the index.ts 1st line.

Upon researching a solution for this error, the majority of sources suggest importing reflect-metadata at the root level file, which I have already done. I wrote Jest test cases for the Injectable class and the test cases are passing when I run them in the terminal, but when I run the extension in the browser, I get the above error. I’m uncertain about what I might be doing incorrectly.

**Comment by PodaruDragos** [comment]
there could be more than 1 reasons for this one. I assume you have the `tsconfig` `emitMetadata` and you are using at least `4.5` version, right ?

**Reply by manoj30075** [reply]
Yes, I have those configurations set in tsconfig and using version 5.2.2. The strange thing that I found out while debugging is that I created another Jupyter extension just to test InversifyJS. I added a simple injectable class. When I installed that sample extension, my main extension started working fine, but the sample extension threw this error in the browser console: Error: Missing required @injectable annotation in: DependencyInjector. Then, when I uninstalled the sample extension, the old error returned for the main extension. Now I'm not sure what's the issue here.

**Reply by femisystems** [reply]
@manoj30075 were you eventually able to fix this problem? I also have a similar problem right now with Jupyterlab 4. With Jupyterlab3, it works well.

**Reply by notaphplover** [reply]
Hey, just open a new issue with a minimum recreation code and I'll have a look at it

**Reply by hi-reeve** [reply]
got this error as well

**Reply by notaphplover** [reply]
@hi-reeve as said, please open a new issue and provide a reproduction repository or a way to easilly recreate the issue. We cannot fix what we cannot recreate

**Reply by hi-reeve** [reply]
> @hi-reeve as said, please open a new issue and provide a reproduction repository or a way to easilly recreate the issue. We cannot fix what we cannot recreate

i havent had time to create repro, but i found out my problem. so if i use this [library](https://github.com/abraham/reflection)  to replace reflect-metadata i got this error because this library does not have the getOwnMetaDataKeys. if i go back using reflect-metadata. it is working fine

**Reply by notaphplover** [reply]
I see, then your issue is not related to inversify for its root cause is using an incomplete polyfill.
